### PR TITLE
refactor: refine pmap with nolink_apply

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

File: apps/emqx/src/emqx_misc.erl [hunk 1]
```diff
@@ -374,10 +380,95 @@ explain_posix(estale) -> "Stale remote file handle";
...
+                    %% For whatever reason, if the caller is dead,
+                    %% there is no reason to continue
+                    exit(Worker, kill),
+                    exit(normal);
+                {'EXIT', Worker, normal} ->
+                    exit(normal);
+                {'EXIT', Worker, Reason} ->
+                    %% worker has finished its job (Reason=normal)
+                    %% or might have crashed due to exception when evaluating Fun
+                    _ = erlang:sendd(Caller, {ResRef, {'EXIT', Reason}}),
```
Comment (thalesmg) [inline apps/emqx/src/emqx_misc.erl:422]:
```suggestion
                    _ = erlang:send(Caller, {ResRef, {'EXIT', Reason}}),
```

File: apps/emqx/src/emqx_misc.erl [hunk 2]
```diff
@@ -374,10 +380,95 @@ explain_posix(estale) -> "Stale remote file handle";
...
+            receive
+                {'DOWN', CallerMRef, process, _, _} ->
+                    %% For whatever reason, if the caller is dead,
+                    %% there is no reason to continue
+                    exit(Worker, kill),
+                    exit(normal);
+                {'EXIT', Worker, normal} ->
+                    exit(normal);
+                {'EXIT', Worker, Reason} ->
+                    %% worker has finished its job (Reason=normal)
```
Comment (thalesmg) [inline apps/emqx/src/emqx_misc.erl:420]: Nit: not the case here, due to the above clause.

File: apps/emqx/src/emqx_misc.erl [hunk 3]
```diff
@@ -375,6 +382,8 @@ explain_posix(estale) -> "Stale remote file handle";
 explain_posix(exdev) -> "Cross-domain link";
 explain_posix(NotPosix) -> NotPosix.
 
+%% @doc Like lists:map/2, only the callback function is evaluated
+%% concurrently.
 -spec pmap(fun((A) -> B), list(A)) -> list(B | {error, term()}).
```
Comment (savonarola) [inline apps/emqx/src/emqx_misc.erl:387]:
```suggestion
-spec pmap(fun((A) -> B), list(A)) -> list(B) | no_return().
```

File: apps/emqx/src/emqx_misc.erl [hunk 4]
```diff
@@ -383,21 +392,90 @@ pmap(Fun, List) when is_function(Fun, 1), is_list(List) ->
 pmap(Fun, List, Timeout) when
```
Comment (savonarola) [inline apps/emqx/src/emqx_misc.erl:392]: Here we should also probably change spec to
```erlang
-spec pmap(fun((A) -> B), list(A), timeout()) -> list(B) | no_return().
```

**Review by ieQu1 — APPROVED** [review]

**Pull request merged**